In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline
from sklearn.metrics import mean_absolute_error,mean_squared_error,r2_score

In [2]:
final=pd.read_csv(r"C:\Users\HP\Downloads\Crop-yield-prediction-using-weather-data-and-NDVI-time-series-main\Phase3\csv\phase3_input_prod.csv")

In [3]:
final

Year  Precipitation  maxtemp  mintemp  avgtemp  specifichumidity  \
0   1999         900.36    32.27    19.14    13.13          0.011637   
1   2000         600.34    32.93    18.87    14.06          0.010941   
2   2001         547.98    32.79    19.17    13.62          0.011391   
3   2002         500.32    33.53    19.61    13.91          0.011091   
4   2003         402.86    33.77    19.89    13.87          0.011109   
5   2004         690.69    32.54    19.13    13.40          0.011452   
6   2005         904.80    31.97    19.00    12.98          0.011786   
7   2006         933.51    31.99    19.09    12.90          0.011998   
8   2007         715.04    32.52    19.43    13.09          0.011772   
9   2008         645.33    32.27    19.14    13.13          0.011637   
10  2009         731.75    32.86    19.80    13.06          0.011776   
11  2010         799.15    32.39    20.20    12.19          0.012701   
12  2011         659.16    32.47    19.16    13.31          0.011652   
13  2012         464.38    33.31    19.47    13.84          0.011018   
14  2013         659.18    32.54    19.66    12.87          0.011902   
15  2014         622.83    32.74    19.69    13.05          0.011856   
16  2015         496.85    33.33    19.83    13.50          0.011725   
17  2016         801.98    32.84    19.31    13.53          0.011382   
18  2017         827.51    32.74    19.21    13.53          0.011535   
19  2018         353.50    33.03    19.62    13.41          0.011057   
20  2019        1057.54    32.80    19.20    13.60          0.011498   

    relativehumidity  surfacepressure  dewpoints  minwindspeed  maxwindspeed  \
0              55.00            94.02      13.99          2.16          5.23   
1              50.60            93.98      12.46          2.58          5.96   
2              53.15            94.03      13.59          2.71          6.20   
3              49.82            94.08      13.21          2.59          6.08   
4              49.22            94.07      13.33          2.55          6.10   
5              53.82            94.09      13.56          2.62          6.02   
6              56.60            94.06      14.19          2.57          6.03   
7              57.74            94.07      14.56          2.58          6.12   
8              55.07            94.02      14.39          2.45          5.96   
9              55.00            94.02      13.99          2.56          6.00   
10             54.63            94.04      14.31          2.54          5.96   
11             59.06            94.01      15.74          2.39          5.74   
12             54.65            94.01      14.11          2.61          6.26   
13             50.10            94.02      12.90          2.61          6.20   
14             55.52            94.02      14.47          2.66          6.36   
15             54.38            94.11      14.81          2.47          6.05   
16             51.94            94.15      14.38          2.37          5.85   
17             53.13            94.10      13.68          2.48          5.91   
18             53.82            94.08      13.67          2.33          5.78   
19             50.04            94.09      13.36          2.29          5.93   
20             53.60            94.10      13.60          2.30          5.80   

    cloudcoverage  Nitrogen  Phosphorus  Pottasium  pH      NDVI    Production  
0            0.55         0           1          1   2  0.369000  1.543140e+06  
1            0.56        -1           1          2   2  0.355333  1.335280e+06  
2            0.56         0           1          1   2  0.403000  1.380978e+06  
3            0.56         0           1          0   2  0.412200  1.358112e+06  
4            0.56        -2          -1          0   2  0.308000  9.820684e+05  
5            0.56        -2           0          1   2  0.458000  1.179750e+06  
6            0.54        -1           1          2   2  0.454000  1.295850e+06  
7         

In [4]:
final.shape

(21, 18)

In [5]:
final.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21 entries, 0 to 20
Data columns (total 18 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Year              21 non-null     int64  
 1   Precipitation     21 non-null     float64
 2   maxtemp           21 non-null     float64
 3   mintemp           21 non-null     float64
 4   avgtemp           21 non-null     float64
 5   specifichumidity  21 non-null     float64
 6   relativehumidity  21 non-null     float64
 7   surfacepressure   21 non-null     float64
 8   dewpoints         21 non-null     float64
 9   minwindspeed      21 non-null     float64
 10  maxwindspeed      21 non-null     float64
 11  cloudcoverage     21 non-null     float64
 12  Nitrogen          21 non-null     int64  
 13  Phosphorus        21 non-null     int64  
 14  Pottasium         21 non-null     int64  
 15  pH                21 non-null     int64  
 16  NDVI              21 non-null     float64
 17 

In [6]:
#Creating correlation for precipitation
#Creating the correlation matrix for the features
corr_matrix=final.corr()
corr_matrix

Year  Precipitation   maxtemp   mintemp   avgtemp  \
Year              1.000000       0.068549  0.095719  0.285614 -0.122287   
Precipitation     0.068549       1.000000 -0.703647 -0.389650 -0.444547   
maxtemp           0.095719      -0.703647  1.000000  0.460998  0.706454   
mintemp           0.285614      -0.389650  0.460998  1.000000 -0.302334   
avgtemp          -0.122287      -0.444547  0.706454 -0.302334  1.000000   
specifichumidity  0.136608       0.521334 -0.628054  0.326134 -0.934561   
relativehumidity  0.014026       0.718991 -0.874397 -0.053217 -0.896917   
surfacepressure   0.482048      -0.028625  0.308355  0.210919  0.163847   
dewpoints         0.202048       0.376681 -0.519871  0.457773 -0.923092   
minwindspeed     -0.354845      -0.294474  0.034467 -0.148253  0.152297   
maxwindspeed      0.049339      -0.414889  0.160301  0.022141  0.152379   
cloudcoverage     0.138877      -0.069324  0.217686 -0.044933  0.270522   
Nitrogen         -0.052433       0.168312 -0.035752  0.104737 -0.121529   
Phosphorus        0.146814       0.382121 -0.179548 -0.066999 -0.138434   
Pottasium         0.242752       0.503336 -0.381170 -0.000657 -0.405969   
pH                0.332347      -0.406142  0.139110  0.138228  0.040769   
NDVI              0.410227       0.167126 -0.423249  0.106024 -0.539005   
Production        0.227175       0.443516 -0.164562 -0.147623 -0.057830   

                  specifichumidity  relativehumidity  surfacepressure  \
Year                      0.136608          0.014026         0.482048   
Precipitation             0.521334          0.718991        -0.028625   
maxtemp                  -0.628054         -0.874397         0.308355   
mintemp                   0.326134         -0.053217         0.210919   
avgtemp                  -0.934561         -0.896917         0.163847   
specifichumidity          1.000000          0.912278        -0.073315   
relativehumidity          0.912278          1.000000        -0.219638   
surfacepressure          -0.073315         -0.219638         1.000000   
dewpoints                 0.967198          0.823876         0.048469   
minwindspeed             -0.137985         -0.052952        -0.298827   
maxwindspeed             -0.160996         -0.153699        -0.093043   
cloudcoverage            -0.258524         -0.242201        -0.209530   
Nitrogen                  0.218598          0.177338        -0.274785   
Phosphorus                0.271946          0.293061        -0.153620   
Pottasium                 0.522805          0.514579        -0.063745   
pH                       -0.287170         -0.314657         0.182857   
NDVI                      0.472440          0.479362         0.111646   
Production                0.195191          0.235817         0.034315   

                  dewpoints  minwindspeed  maxwindspeed  cloudcoverage  \
Year               0.202048     -0.354845      0.049339       0.138877   
Precipitation      0.376681     -0.294474     -0.414889      -0.069324   
maxtemp           -0.519871      0.034467      0.160301       0.217686   
mintemp            0.457773     -0.148253      0.022141      -0.044933   
avgtemp           -0.923092      0.152297      0.152379       0.270522   
specifichumidity   0.967198     -0.137985     -0.160996      -0.258524   
relativehumidity   0.823876     -0.052952     -0.153699      -0.242201   
surfacepressure    0.048469     -0.298827     -0.093043      -0.209530   
dewpoints          1.000000     -0.167377     -0.121884      -0.315324   
minwindspeed      -0.167377      1.000000      0.866176       0.042914   
maxwindspeed      -0.121884      0.866176      1.000000       0.057430   
cloudcoverage     -0.315324      0.042914      0.057430       1.000000   
Nitrogen           0.194936      0.003493     -0.096450      -0.136837   
Phosphorus         0.204297      0.082957     -0.011011      -0.183562   
Pottasium          0.447673     -0.045022     -0.098693      -0.172075   
pH                -0.178

In [8]:
mask = np.zeros_like(corr_matrix, dtype=bool)
mask[np.triu_indices_from(mask)]= True

In [9]:
f, ax = plt.subplots(figsize=(18, 15)) 
heatmap = sns.heatmap(corr_matrix, mask = mask,square = True,linewidths = .5, cmap = 'coolwarm', cbar_kws = {'shrink': .4, 'ticks' : [-1, -.5, 0, 0.5, 1]},vmin = -1, vmax = 1,annot = True,annot_kws = {'size': 12})
#add the column names as labels
ax.set_yticklabels(corr_matrix.columns, rotation = 0)
ax.set_xticklabels(corr_matrix.columns)
sns.set_style({'xtick.bottom': True}, {'ytick.left': True})

In [10]:
#final=final.drop(['maxtemp','mintemp', 'avgtemp','pH','cloudcoverage','maxwindspeed','minwindspeed'],axis=1)

In [11]:
final.shape

(21, 18)

In [12]:
final1=final.iloc[:,:].values

In [13]:
final1

array([[ 1.99900000e+03,  9.00360000e+02,  3.22700000e+01,
         1.91400000e+01,  1.31300000e+01,  1.16370000e-02,
         5.50000000e+01,  9.40200000e+01,  1.39900000e+01,
         2.16000000e+00,  5.23000000e+00,  5.50000000e-01,
         0.00000000e+00,  1.00000000e+00,  1.00000000e+00,
         2.00000000e+00,  3.69000000e-01,  1.54313992e+06],
       [ 2.00000000e+03,  6.00340000e+02,  3.29300000e+01,
         1.88700000e+01,  1.40600000e+01,  1.09410000e-02,
         5.06000000e+01,  9.39800000e+01,  1.24600000e+01,
         2.58000000e+00,  5.96000000e+00,  5.60000000e-01,
        -1.00000000e+00,  1.00000000e+00,  2.00000000e+00,
         2.00000000e+00,  3.55333333e-01,  1.33528017e+06],
       [ 2.00100000e+03,  5.47980000e+02,  3.27900000e+01,
         1.91700000e+01,  1.36200000e+01,  1.13910000e-02,
         5.31500000e+01,  9.40300000e+01,  1.35900000e+01,
         2.71000000e+00,  6.20000000e+00,  5.60000000e-01,
         0.00000000e+00,  1.00000000e+00,  1.00000000e

In [14]:
X=final1[:,:-1]
y=final1[:,-1]

In [15]:
X

array([[ 1.99900000e+03,  9.00360000e+02,  3.22700000e+01,
         1.91400000e+01,  1.31300000e+01,  1.16370000e-02,
         5.50000000e+01,  9.40200000e+01,  1.39900000e+01,
         2.16000000e+00,  5.23000000e+00,  5.50000000e-01,
         0.00000000e+00,  1.00000000e+00,  1.00000000e+00,
         2.00000000e+00,  3.69000000e-01],
       [ 2.00000000e+03,  6.00340000e+02,  3.29300000e+01,
         1.88700000e+01,  1.40600000e+01,  1.09410000e-02,
         5.06000000e+01,  9.39800000e+01,  1.24600000e+01,
         2.58000000e+00,  5.96000000e+00,  5.60000000e-01,
        -1.00000000e+00,  1.00000000e+00,  2.00000000e+00,
         2.00000000e+00,  3.55333333e-01],
       [ 2.00100000e+03,  5.47980000e+02,  3.27900000e+01,
         1.91700000e+01,  1.36200000e+01,  1.13910000e-02,
         5.31500000e+01,  9.40300000e+01,  1.35900000e+01,
         2.71000000e+00,  6.20000000e+00,  5.60000000e-01,
         0.00000000e+00,  1.00000000e+00,  1.00000000e+00,
         2.00000000e+00,  4.0

In [16]:
y

array([1543139.923 , 1335280.167 , 1380977.99  , 1358112.463 ,
        982068.3741, 1179749.613 , 1295849.778 , 1221359.357 ,
       1232805.848 , 1224945.893 , 1305474.725 , 1373398.91  ,
       1439374.55  , 1361322.563 , 1468101.096 , 1444372.796 ,
       1410443.725 , 1815274.444 , 1444922.863 ,  982068.3741,
       1444922.863 ])

In [17]:
X.shape

(21, 17)

In [18]:
y.shape

(21,)

In [19]:
#train test split
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.1,random_state=101)

In [20]:
X_train,y_train=np.array(X_train),np.array(y_train)

    ### Decision Tree Regressor

In [21]:
#Decision Tree Regressor
from sklearn.tree import DecisionTreeRegressor
dt=DecisionTreeRegressor()

In [22]:
dt.fit(X_train,y_train)

DecisionTreeRegressor()

In [23]:
dt_ypred= dt.predict(X_test)

In [24]:
print('Mean squared Error:',mean_squared_error(y_test,dt_ypred))
print('Root Mean Squared Error:',np.sqrt(mean_squared_error(y_test,dt_ypred)))
print('R2_Score:',r2_score(y_test,dt_ypred))

Mean squared Error: 2477371273.8019204
Root Mean Squared Error: 49773.19834812628
R2_Score: 0.7197110339367045


### Random Forest Regressor

In [25]:
from sklearn.ensemble import RandomForestRegressor

In [26]:
rf = RandomForestRegressor(n_estimators = 1000, random_state = 42)
rf.fit(X_train, y_train);

In [27]:
rf_pred=rf.predict(X_test)

In [28]:
print('Mean squared Error:',mean_squared_error(y_test,rf_pred))
print('Root Mean Squared Error:',np.sqrt(mean_squared_error(y_test,rf_pred)))
print('R2_Score:',r2_score(y_test,rf_pred))

Mean squared Error: 672491666.2017423
Root Mean Squared Error: 25932.444277424802
R2_Score: 0.9239145154385365


# Random Forest Regressor is best for production prediction with accuracy of 92.21%.

In [31]:
prod2020=pd.read_csv(r"C:\Users\HP\Downloads\Crop-yield-prediction-using-weather-data-and-NDVI-time-series-main\Phase3\csv\2020prediction.csv")

In [32]:
prod2020=prod2020.iloc[:,:].values

In [33]:
prod2020

array([[ 2.02000000e+03,  7.45340000e+02,  3.27565278e+01,
         1.93041667e+01,  1.35251191e+01,  1.13656250e-02,
         5.05465100e+01,  9.40862964e+01,  1.34754762e+01,
         2.45597222e+00,  5.95358333e+00,  5.57666667e-01,
        -1.00000000e+00,  1.00000000e+00,  2.00000000e+00,
         2.00000000e+00,  4.20000000e-01]])

In [34]:
prod2020_pred=[]

In [37]:
whos


Variable                Type                     Data/Info
----------------------------------------------------------
DecisionTreeRegressor   ABCMeta                  <class 'sklearn.tree._cla<...>s.DecisionTreeRegressor'>
RandomForestRegressor   ABCMeta                  <class 'sklearn.ensemble.<...>t.RandomForestRegressor'>
X                       ndarray                  21x17: 357 elems, type `float64`, 2856 bytes
X_test                  ndarray                  3x17: 51 elems, type `float64`, 408 bytes
X_train                 ndarray                  18x17: 306 elems, type `float64`, 2448 bytes
ax                      Axes                     Axes(0.125,0.123;0.62x0.744)
corr_matrix             DataFrame                                      Yea<...>ction          1.000000  
dt                      DecisionTreeRegressor    DecisionTreeRegressor()
dt_ypred                ndarray                  3: 3 elems, type `float64`, 24 bytes
f                       Figure                   Fi

In [44]:
prod2020_pred=dt.predict(prod2020)


In [45]:
prod2020_pred

array([1410443.725])

# The overall sugarcane production for Vedhani for year 2020 is 1410443.725 tonnes.